### PR TITLE
Improvements03

### PR DESCRIPTION
**Pull request by sungysang**
FYI I deleted the original trigger branch and created a new one since I messed that one up. This is the second draft with added recommendations from Josiah. Flow is:

Lines 37-45: Create a Set(zipCodeSet) of zipcodes from all accounts that initiate the trigger. Filters added to check if the Parent account is empty and record type is church. <-- Change ParentId to districtlookup?

Lines 47-50: Create a List(matchedSchoolDistricts) of Accounts whose zip code matches that of the ones in zipCodeSet. Filters added to check if record type is School District.

Line 52: Create Map (zipCodeMap) to add the zip code and Accounts from matchedSchoolDistricts to use to match to the original accounts.

Line 53: Create empty set (zipCodeKeyDuplicates) to place duplicate zip codes inside for removal.

Lines 54-62: Use a for loop to add each Account into the Map. Using an if statement to add a zip code into the duplicates list if it's already in the Map.

Lines 64-66: Another for loop to remove all the duplicate zip codes from the map.

Lines 77-81: Last for loop to check if the original Account's zip code is inside the map of matches codes. If there's a match, the Account that corresponds to the zipcode is added to the original Account.

There's still a TON of improvements that can be made but I think these are a step towards following best practice guidelines and believe this should now be able to handle Data Loader. Still have not tested this in the Sandbox either.

Should lines 52 and 53 be initialized in the beginning of the trigger? Doesn't really affect the trigger much but wondering if this would be better in terms of best practice.

Created a version of this in Anonymous Apex and it works as intended in Sandbox (IF THE SCHOOLDISTRICT LOOKUP FIELD IS USED INSTEAD OF PARENTID), can begin testing whenever

### CODE REVIEW

**Comment by Julian88Tex** [comment]
@sungysang for the comments on this PR, mind adding them to the actual lines in the file changes section? I'll add a sample comment to show you. It will be easier to review if you want us to look at specific. You can also tag people you specifically want to look at certain lines. I really think David Young would be another good set of eyes on this and I suspect he'd be happy to take a look.

File: account-trigger-zip-code.cls [hunk 1]
```diff
@@ -17,3 +18,67 @@ trigger populateSchoolDistrictBasedOnZip on Account (before update, before inser
...
+
+2. Run a soql to find all matched Parent Accounts with same zip codes 
+
+3. Create a map of Account with zip code as key
+
+4. Populate the Account lookup with the key value.
+
+    Set<String> zipCodeSet = new Set<String>();
+
+    for (Account a : [SELECT BillingPostalCode
```
Comment (Julian88Tex) [inline account-trigger-zip-code.cls:37]: @sungysang your sample comment:

> Create a Set(zipCodeSet) of zipcodes from all accounts that initiate the trigger. Filters added to check if the Parent account is empty and record type is church. <-- Change ParentId to districtlookup?

File: account-trigger-zip-code.cls [hunk 2]
```diff
@@ -17,3 +18,67 @@ trigger populateSchoolDistrictBasedOnZip on Account (before update, before inser
...
+                        FROM Account
+                        WHERE RecordTypeId = '0121N000000qpORQAY'
+                        AND Id IN :Trigger.new
+                        AND ParentId = null //ParentId or DistrictLookupId?
+                        AND BillingPostalCode != null]) {
+
+                            zipCodeSet.add(a.BillingPostalCode);
+                        }
+    
+    List<Account> matchedSchoolDistricts = new List<Account>([SELECT Id, BillingPostalCode
```
Comment (sungysang) [inline account-trigger-zip-code.cls:47]: Create a List(matchedSchoolDistricts) of Accounts whose zip code matches that of the ones in zipCodeSet. Filters added to check if record type is School District.

File: account-trigger-zip-code.cls [hunk 3]
```diff
@@ -17,3 +18,67 @@ trigger populateSchoolDistrictBasedOnZip on Account (before update, before inser
...
+                                            FROM Account
+                                            WHERE Type = 'School District' 
+                                            AND BillingPostalCode IN :zipCodeSet]);
+
+    Map<String, Account> zipCodeMap = new Map<String, Account> ();
```
Comment (sungysang) [inline account-trigger-zip-code.cls:52]: Create Map (zipCodeMap) to add the zip code and Accounts from matchedSchoolDistricts to use to match to the original accounts.

File: account-trigger-zip-code.cls [hunk 4]
```diff
@@ -17,3 +18,67 @@ trigger populateSchoolDistrictBasedOnZip on Account (before update, before inser
...
+    Set<String> zipCodeKeyDuplicates = new Set<String>();
```
Comment (sungysang) [inline account-trigger-zip-code.cls:53]: Create empty set (zipCodeKeyDuplicates) to place duplicate zip codes inside for removal.

File: account-trigger-zip-code.cls [hunk 5]
```diff
@@ -17,3 +18,67 @@ trigger populateSchoolDistrictBasedOnZip on Account (before update, before inser
...
+
+4. Populate the Account lookup with the key value.
+
+    Set<String> zipCodeSet = new Set<String>();
+
+    for (Account a : [SELECT BillingPostalCode
+                        FROM Account
+                        WHERE RecordTypeId = '0121N000000qpORQAY'
+                        AND Id IN :Trigger.new
+                        AND ParentId = null //ParentId or DistrictLookupId?
```
Comment (sungysang) [inline account-trigger-zip-code.cls]: Got confirmation that it should be the District Lookup field

File: account-trigger-zip-code.cls [hunk 6]
```diff
@@ -17,3 +18,67 @@ trigger populateSchoolDistrictBasedOnZip on Account (before update, before inser
...
+    Set<String> zipCodeKeyDuplicates = new Set<String>();
+    for (Account p : matchedSchoolDistricts) {
+
+        if (zipCodeMap.containsKey(p.BillingPostalCode)) {
+            zipCodeKeyDuplicates.add(p.BillingPostalCode);
+        }
+        else {
+            zipCodeMap.put(p.BillingPostalCode, p);
+        }
+    }
```
Comment (sungysang) [inline account-trigger-zip-code.cls:62]: Use a for loop to add each Account into the Map. Using an if statement to add a zip code into the duplicates list if it's already in the Map.

File: account-trigger-zip-code.cls [hunk 7]
```diff
@@ -17,3 +18,67 @@ trigger populateSchoolDistrictBasedOnZip on Account (before update, before inser
...
+
+    for (String zipCode : zipCodeKeyDuplicates) {
+        zipCodeMap.remove(zipCode);
+    }
```
Comment (sungysang) [inline account-trigger-zip-code.cls:66]: Another for loop to remove all the duplicate zip codes from the map.

File: account-trigger-zip-code.cls [hunk 8]
```diff
@@ -17,3 +18,67 @@ trigger populateSchoolDistrictBasedOnZip on Account (before update, before inser
...
+        // loop through this keylist and use it to remove all keys from the map.
+        // if contains key ==> add to keylist ==> for string key : keylist {map.remove(key)}
+        // COMPLETED
+        
+    for (account a : Trigger.New) {
+        if (zipCodeMap.containsKey(a.BillingPostalCode)) {
+            Account match = zipCodeMap.get(a.BillingPostalCode);
+            a.ParentId = match.Id;
+        }
+    }
```
Comment (sungysang) [inline account-trigger-zip-code.cls:82]: Last for loop to check if the original Account's zip code is inside the map of matches codes. If there's a match, the Account that corresponds to the zipcode is added to the original Account.

**Comment by Julian88Tex** [comment]
@sungysang Thanks! Btw, if you need to update code on a PR you can just push commits to the same branch. No need to close and recreate a PR.

**Comment by Julian88Tex** [comment]
@sungysang mind making me an admin for the repo? I can add Josiah & David.

**Comment by sungysang** [comment]
@Julian88Tex Sure, but I've already sent a collab invite out to Josiah and have messaged David to see if he's willing to take a look

**Comment by imaliveagain** [comment]
trigger populateSchoolDistrictBasedOnZip on Account (before update, before insert) {
    
    // First loop through all the accounts being passed in the Trigger.new to identify if it's a church and it has zip add it to a list or map
    // declared outside the loop
    // The most efficent way to do this to place the accounts in a map with the zip being the index.
    // because you may have multiple accounts with the same zip you will need to put it in a map like this Map<String,List<Account>>()
    // the below is a good design pattern for populating a index map of lists when you may have multiple objects with the same index key
    /**********
    Map<String,List<Account>> churchMap = new Map<String,List<Account>>();

    for(Account a : Trigger.new){
        // Do you want change a district if it's already populated?
        if(a.Type == 'Church' && a.BillingPostalCode != null && parentId == null){
            if(churchMap.containsKey(a.BillingPostalCode)){
                churchMap.get(a.BillingPostalCode).add(a);
            }else{
                List<Account> accts = new List<Account>();
                accts.add(a);
                churchMap.put(a.BillingPostalCode,accts);
            }

        }
    }
    ************/
    // Now run your SOQL on the mapKeyset to return schools with the churches zip code 
    /***********
     List<Account> matchedSchoolDistricts = [Select Id,Type,BillingPostalCode From Account Where BillingPostalCode =: churchMap.keySet() AND Type = 'Church'];
    ***********/
     // loop through the matched school districts and check if there are multiple school accounts returned for that zip and to create a new
     // of valid zip for that church
     /*******************
     Set<String> existingZips = new Set<String>();
     Map<String,String> noDuplicateZips = new Map<String,String>();
     for(Account a: matchedSchooldistricts){
        if(existingZips.contains(a.BillingPostalCode)){
            noDuplicateZips.remove(a.BillingPostalCode);
        }else{
            existingZips.add(a.BillingPostalCode)
            noDuplicateZips.put(a.BillingPostalCode,a.Id);
        }
     }
     *******************/

     //now you have validated that you are not using a postal code that returned two valid school districts so loop through the Trigger.New
     // and use the no duplicate match
     /*************

     for(Account a: Trigger.new){
        if(a.Type == 'Church' && a.BillingPostalCode != null && parentId == null && noDuplicateZips.containsKey(a.BillingPostalCode)){
            a.parentId = noDuplicateZips.get(a.BillingPostalCode);
        }
     }

     *************/


     /******************  Old Code
    for (Account a : Trigger.new) {

        String schoolZip = a.BillingPostalCode;
        String schoolType = a.Type;
        if (a.parentid != null || schoolType != 'School' || schoolZip == null) {
              return;
        }
    		else {
                List<Account> matchedSchooldistricts = [Select Id from Account where Type = 'School District' and BillingPostalCode != null and BillingPostalCode = :schoolZip];
                    if (matchedSchooldistricts.size() > 1 || matchedSchooldistricts.isEmpty()) {
                        return;    
                    }
                    else {
                        a.parentId = matchedSchooldistricts[0].id;
                    }
            }
    }
     Old Code *******************/
}

**Comment by Julian88Tex** [comment]
@imaliveagain is that a comment or a rewrite?

**Comment by imaliveagain** [comment]
@Julian88Tex Well it's "commented out". Look I don't get to code much with easy stuff so it was a fun 45 min break. Plus @sungysang could benefit from some bulkified design patterns. You could upload 50k accounts without a single issue from this trigger like this.

**Comment by Julian88Tex** [comment]
@imaliveagain Oh right, I forgot your code doesn't have bugs.

**Comment by imaliveagain** [comment]
@Julian88Tex if you look at the acceptance criteria you see the trigger is supposed to act off church accounts not when school district accounts are inserted or updated. Plus the cardinal rule of putting a soql query inside a for loop was committed. When starting out It makes more sense to me to throw out a proper design pattern then explain why it works.

**Comment by Julian88Tex** [comment]
@imaliveagain 🙇

**Comment by imaliveagain** [comment]
> @imaliveagain Oh right, I forgot your code doesn't have bugs.

@Julian88Tex I forgot you don't let code get by without having to comment about it.

**Comment by Julian88Tex** [comment]
@imaliveagain some things don't change ;-)

**Comment by imaliveagain** [comment]
> @imaliveagain some things don't change ;-)
FYI I mistakenly looked at the original code rather than his updates haha so most of it was good.

**Pull request merged**